contrib/dlb/source/connection_balancer_impl.cc

```diff
@@ -0,0 +1,301 @@
+  if (num_rx == 0) {
+    ENVOY_LOG(debug, "{} dlb receive none, skip", name_);
+    return;
+  } else {
+    ENVOY_LOG(debug, "{} get dlb event {}", name_, num_rx);
+  }
+
+  int ret = dlb_release(DlbConnectionBalanceFactorySingleton::get().rx_ports.at(index_), num_rx);
+  if (ret != num_rx) {
+    ENVOY_LOG(debug, "{} dlb release {}", name_, errorDetails(errno));
+  }
+  for (int i = 0; i < num_rx; i++) {
+    if (dlb_events[i].recv.error) {
+      ENVOY_LOG(error, "{} dlb receive {}", name_, errorDetails(errno));
+      continue;
+    }
+
+    if (dlb_events[i].recv.udata64) {
+      const uint64_t data = dlb_events[i].recv.udata64;
+      auto socket = reinterpret_cast<Network::ConnectionSocket*>(data);
+
+      ENVOY_LOG(debug, "{} dlb recv {}", name_, socket->ioHandle().fdDoNotUse());
+      auto listener = dynamic_cast<Envoy::Server::ActiveTcpListener*>(&handler_);
+      listener->onAcceptWorker(std::unique_ptr<Network::ConnectionSocket>(socket),
+                               listener->config_->handOffRestoredDestinationConnections(), true);
+      listener->incNumConnections();
+    }
+  }
+}
+
+void DlbConnectionBalancerImpl::registerHandler(
+    Envoy::Network::BalancedConnectionHandler& handler) {
+
+  auto listener = dynamic_cast<Envoy::Server::ActiveTcpListener*>(&handler);
+
+  const std::string worker_name = listener->dispatcher().name();
+  const int index =
+      std::stoi(worker_name.substr(worker_name.find_first_of('_') + 1, worker_name.size()));
+
+  auto dlb_handler =
+      std::make_shared<DlbBalancedConnectionHandlerImpl>(handler, index, worker_name);
+  dlb_handler->setDlbEvent();
+
+  absl::MutexLock lock(&DlbConnectionBalanceFactorySingleton::get().lock);
+  DlbConnectionBalanceFactorySingleton::get().dlb_handlers.push_back(dlb_handler);
+}
+
+void DlbConnectionBalancerImpl::unregisterHandler(
+    Envoy::Network::BalancedConnectionHandler& handler) {
+  auto listener = dynamic_cast<Envoy::Server::ActiveTcpListener*>(&handler);
+  auto worker_name = listener->dispatcher().name();
+  int index = std::stoi(worker_name.substr(worker_name.find_first_of('_') + 1, worker_name.size()));
+
+  // Now DLB does not support change config when running, clean DLB related config in
+  // DlbBalancedConnectionHandlerImpl
+  absl::MutexLock lock(&DlbConnectionBalanceFactorySingleton::get().lock);
+  auto dlb_handlers = DlbConnectionBalanceFactorySingleton::get().dlb_handlers;
+  dlb_handlers.erase(dlb_handlers.begin() + index);
+}
+
+Envoy::Network::BalancedConnectionHandler& DlbConnectionBalancerImpl::pickTargetHandler(
+    Envoy::Network::BalancedConnectionHandler& current_handler) {
+  auto listener = dynamic_cast<Envoy::Server::ActiveTcpListener*>(&current_handler);
+  auto worker_name = listener->dispatcher().name();
+  const int index =
+      std::stoi(worker_name.substr(worker_name.find_first_of('_') + 1, worker_name.size()));
+  absl::MutexLock lock(&DlbConnectionBalanceFactorySingleton::get().lock);
```

Comment:
you are using the lock again, that will be the same as `ExactConnectionBalancer` getting bad throughput I guess.

Reply:
All tests without lock at local pass, remove it :)